### PR TITLE
Fix example ERC20 contract

### LINKED ISSUE #1095: Tests of ERC20 Solidity compatible doesn't work
### Version Information

* vyper Version: 0.1.0b4
* OS: osx
* Python Version: 3.6.3

### What's your issue about?

I'm going to update ERC20 in `example` repository to the latest Vyper version (`0.1.0b4`) and working on the test of this. I tried the current tests in [tests/examples/tokens/ERC20_solidity_compatible](https://github.com/ethereum/vyper/tree/master/tests/examples/tokens/ERC20_solidity_compatible) with `python run_tests.py ` command but it causes this error.

```
(vyper) nrryuya:ERC20_solidity_compatible nrryuya1$ python run_tests.py 

Failed to import bitcoin. This is not a fatal error but does
mean that you will not be able to determine the address from
your wallet file.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Running Solidity-compatible ERC20 tests, Suite 1 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Initializing chain from provided state
Initializing chain from provided state
ERROR
Initializing chain from provided state
Initializing chain from provided state
test_deposit_and_withdraw (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_failed_send_in_withdraw (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_initial_state (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_maxInts (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_payability (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_raw_logs (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_totalSupply (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_transfer (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
test_transferFromAndAllowance (test.erc20_tests_1.TestSolidity1ERC20) ... ERROR
Initializing chain from provided state
Initializing chain from provided state
test_deposit_and_withdraw (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_failed_send_in_withdraw (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_initial_state (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_maxInts (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_payability (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_raw_logs (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_totalSupply (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_transfer (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR
test_transferFromAndAllowance (test.erc20_tests_1.TestSolidity2ERC20) ... ERROR

======================================================================
ERROR: setUpClass (test.erc20_tests_1.TestVyperERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 343, in setUpClass
    cls.t.languages['vyper'] = compiler.Compiler()
AttributeError: module 'vyper.compiler' has no attribute 'Compiler'

======================================================================
ERROR: test_deposit_and_withdraw (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_failed_send_in_withdraw (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_initial_state (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_maxInts (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_payability (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_raw_logs (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_totalSupply (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_transfer (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_transferFromAndAllowance (test.erc20_tests_1.TestSolidity1ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 406, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_deposit_and_withdraw (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_failed_send_in_withdraw (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_initial_state (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_maxInts (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_payability (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_raw_logs (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_totalSupply (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_transfer (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_transferFromAndAllowance (test.erc20_tests_1.TestSolidity2ERC20)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 422, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py", line 52, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

----------------------------------------------------------------------
Ran 18 tests in 0.415s

FAILED (errors=19)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Finished Solidity-compatible ERC20 tests, Suite 1 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Running Solidity-compatible ERC20 tests, Suite 2 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Initializing chain from provided state
Testing nonvyper/ERC20_solidity_2.sol
Deploying contract at 09f24956cac320a987934f5da542b19a4e1ff04e used 371261 gas
test_all (test.erc20_tests_2.ERC20_solidity_2_sol) ... ERROR
Initializing chain from provided state
Testing nonvyper/ERC20_solidity_1.sol
Deploying contract at 09f24956cac320a987934f5da542b19a4e1ff04e used 378559 gas
test_all (test.erc20_tests_2.ERC20_solidity_1_sol) ... ERROR

======================================================================
ERROR: test_all (test.erc20_tests_2.ERC20_solidity_2_sol)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_2.py", line 144, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/utils/pyethereum_test_utils.py", line 54, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

======================================================================
ERROR: test_all (test.erc20_tests_2.ERC20_solidity_1_sol)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_2.py", line 144, in setUp
    super().setUp()
  File "/Users/nrryuya1/work/blockchain/nrryuya/vyper/tests/examples/tokens/ERC20_solidity_compatible/utils/pyethereum_test_utils.py", line 54, in setUp
    self.s.revert(self.initial_state)
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/ethereum/tools/tester.py", line 297, in revert
    self.block.transactions = self.block.transactions[:txcount]
  File "/Users/nrryuya1/.pyenv/versions/vyper/lib/python3.6/site-packages/rlp/sedes/serializable.py", line 334, in field_fn_setter
    raise AttributeError("can't set attribute")
AttributeError: can't set attribute

----------------------------------------------------------------------
Ran 2 tests in 0.287s

FAILED (errors=2)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
Finished Solidity-compatible ERC20 tests, Suite 2 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
FAILED : Test failures encounting, exiting with error.
```

### How can it be fixed?
As the first `AttributeError: module 'vyper.compiler' has no attribute 'Compiler'` shows, `Compiler` object is removed from `vyper/compiler.py` in [this commit](https://github.com/ethereum/vyper/commit/077c72a6e14e8386bffc92229cc37da7afaf1f98). We need to fix this somehow?

EDIT: I added the `Compiler` class in `compiler.py` and `AttributeError: module 'vyper.compiler' has no attribute 'Compiler'` disappeared. However, `AttributeError: can't set attribute` still remains.

#### Cute Animal Picture

image

### PR DESCRIPTION
**Pull request by nrryuya**
### - What I did
Fix #1095.

### - How I did it
- Unified the existing three ERC20 contract (ERC20.vy, vypercoin.vy, ERC20_solidity_compatible) to only one Solidity compatible contract 
  - Removed old ERC20_solidity_compatible repository
- Refactored tests
   - Rewrote [the Philip Daian's tests](https://github.com/ethereum/vyper/blob/v0.1.0-beta.5/tests/examples/tokens/ERC20_solidity_compatible/test/erc20_tests_1.py) with pytest and web3 so that it's included in the CI
  - Removed the tests for ERC20 in Solidity which I guess for the checks of compatibility because it's out of the scope of this example repository. 
- Added mint & burn functions
  - Removed deposit & withdraw functions which allow users to exchange token at the ratio 1 wei = 1 token anytime because it is not a standard implementation

### - How to verify it
tests

### - Description for the changelog
Fix example ERC20 contract

### - Cute Animal Picture

image

### CODE REVIEW

File: tests/examples/tokens/test_erc20.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+    name = bytes_helper("Vypercoin", 32)
+    symbol = bytes_helper("VYP", 32)
+    bad_code = code.replace("self.total_supply = self.total_supply + _value", "") \
+                .replace("self.total_supply = self.total_supply - _value", "")
+    c = get_contract(bad_code, name, symbol, 0, 0)
+    return c
+
+
+@pytest.fixture
+def get_log_args(get_logs):
```
Comment (nrryuya) [inline tests/examples/tokens/test_erc20.py]: We can reuse this fixture in `test_erc721.py` and `test_company.py` so I'll create a new pr after this gets merged and add this fixture to conftest.py.

**Comment by nrryuya** [comment]
Oh vypercoin.vy is used in other two tests. I'll fix this.

**Comment by fubuloubu** [comment]
Shall I close #1034 with this update?

**Comment by nrryuya** [comment]
@fubuloubu 
Yeah! Thank you for the prior work ;)

File: examples/tokens/ERC20.vy [hunk 2]
```diff
@@ -1,87 +1,138 @@
...
-# If this function is called again it overwrites the current allowance with _value.
+#  @dev Transfer tokens from one address to another.
+#       Note that while this function emits an Approval event, this is not required as per the specification,
+#       and other compliant implementations may not emit the event.
+#  @param _from address The address which you want to send tokens from
+#  @param _to address The address which you want to transfer to
+#  @param _value uint256 the amount of tokens to be transferred
 @public
-def approve(_spender: address, _amount: int128(uint256)) -> bool:
+def transferFrom(_from : address, _to : address, _value : uint256) -> bool:
```
Comment (jacqueswww) [inline examples/tokens/ERC20.vy]: isn't  return False still needed, or has that changed? If the if is removed, don't we still need the assertion of balances? (same question for def transfer).

File: tests/examples/tokens/test_erc20.py (hunk 1 above)
Comment (jacqueswww) [reply tests/examples/tokens/test_erc20.py]: Just check `conftest.py` I feel like we have something similar already ;)

Comment (nrryuya) [reply tests/examples/tokens/test_erc20.py]: You mean [`get_logs`](https://github.com/ethereum/vyper/blob/v0.1.0-beta.5/tests/conftest.py#L300)? We extended it to `get_log_args` which extracts the `args` attribute from `logs`.

File: examples/tokens/ERC20.vy (hunk 2 above)
Comment (nrryuya) [reply examples/tokens/ERC20.vy]: It seems that there had been a debate about "return false" vs "throw" but now the EIP is fixed to "throw". From the [EIP](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md#transfer): 
> The function SHOULD throw if the _from account balance does not have enough tokens to spend.
[OpenZeppelin's ERC20 contract](https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/token/ERC20/ERC20.sol#L136) doesn't return false.

 About assertions, this subtraction reverts if the balance of `_from` account is not enough. This is checked in the test.

```
self.balances[_from] = self.balances[_from] - _value
```

**Pull request merged**